scripts/clean-old-namespace-from-cluster.sh

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env sh
+# this scripts logins as kubeadmin and removes the namespaces that are more than oneday old.
+# this script is used in a kubernetes cronjob for PSI/IBM cloud openshift cluster.  
+
+# login as kubeadmin
+if [[ $CLUSTER_TYPE == "PSI" ]]; then
+    ￼ #PSI cluster login
+    ￼ oc login -u kubeadmin -p ${OCP4X_KUBEADMIN_PASSWORD} --insecure-skip-tls-verify ${OCP4X_API_URL}
+else
+    ￼ # Login to IBM Cloud using service account API Key
+    ￼ ibmcloud login --apikey $IBMC_OCP47_APIKEY -a cloud.ibm.com -r eu-de -g "Developer CI and QE"
+    ￼
+    ￼ # Login to cluster in IBM Cloud using cluster API key
+    ￼ oc login --token=$IBMC_OCLOGIN_APIKEY --server=$IBMC_OCP47_SERVER
+fi
+
+# PROJECT_AND_TIME var will contain namespace and date with time seperated with `|`
+# PROJECT_AND_TIME doesn't contain openshift/ibm/kube namespace
+# eg. cmd-push-test157kgb|2021-08-17T12:40:20Z
+PROJECT_AND_TIME=$(kubectl get projects -o jsonpath='{range .items[*]}{.metadata.name}{"|"}{.metadata.creationTimestamp} {"\n"}{end}' | grep -v '^openshift\|^kube\|^default\|^ibm\|^calico\|^tigera')
```

Comment:
I understood we were running on a cluster in common with other teams. Won't we delete other teams namespaces this way?

Reply:
yes we will remove the namespace for other teams too, but this was discussed on the [issue](https://github.com/openshift/odo/issues/4467#issuecomment-893363050) as this will be a common solution for every team.

Reply:
Cluster will be in trouble each time someone in a team will install something in a new namespace. It is not a very robust solution.

I would prefer to use a prefix for each team to prefix the namespaces to delete, and grep on this prefix, instead of invert-grepping on the namespaces to remove.

@rnapoles-rh @kadel WDYT?

Reply:
+1 feloy I suggested the same on this [comment](https://github.com/openshift/odo/issues/4467#issuecomment-891777443)

Reply:
I guess the deletion of projects based on time still stands, and this is more a discussion on which projects will go under the knife in the first place. 

I guess we could have a prefix on projects which can be deleted and only put those projects through the age check before removing them. 

This also means we have to communicate to every team beforehand about this when they are onboarded into the cluster and make it clear that if a prefix is not set to the tests, then it may not get cleaned up resulting in not only leaks in resources, but also usage costs

Reply:
Also, namespaces cannot be renamed, at least not easily, so the name with prefix should be set during the creation

Reply:
> I would prefer to use a prefix for each team to prefix the namespaces to delete, and grep on this prefix, instead of invert-grepping on the namespaces to remove.

This could lead to other problems. For example, we don't have a mechanism to enforce prefixes. So if someone creates namespace without correct prefix, it will be there forever. 

I think that invert-grepping can be ok, but it needs to be clearly communicated to everyone who is using the cluster that this mechanism is there.

Reply:
> This could lead to other problems. For example, we don't have a mechanism to enforce prefixes. So if someone creates namespace without correct prefix, it will be there forever.

but those namespace will be created by someone manually, so it may have some importance to it for that time,
and we can add an additional layer of check for non prefix namespace which are more then 20/30 days old will be deleted?

Reply:
The Hierarchical namespaces (https://kubernetes.io/blog/2020/08/14/introducing-hierarchical-namespaces/) could help resolve these multi-tenancy problems

Reply:
> The Hierarchical namespaces (https://kubernetes.io/blog/2020/08/14/introducing-hierarchical-namespaces/) could help resolve these multi-tenancy problems

But is this something that can be done in the background without significant changes to testing. Also, prevention of deletion of constant namespaces can be handled using an exclusion list akin to gitignore

Basically
```
for ns in namespaces
{
   if ns != kube* || ns != openshift* || ns not in exclusion-list {
     if ns is too old 
     {
         delete it
     }
  }
}
```

Reply:
The exclusion list will be difficult to maintain, and subject to some errors: we will probably remember to add a namespace to the list only after the first time it is deleted, which can be painful.

Reply:
> The exclusion list will be difficult to maintain, and subject to some errors: we will probably remember to add a namespace to the list only after the first time it is deleted, which can be painful.

This is fine, the reverse of this is an improperly named ns being left behind(with all of its resources) which is going to increase operational costs. If a test user forgets to add an ns, then it is his fault, not ours.

I agree that communication should be clear and people should be aware cleanup happens and how to avoid it.
An exclusion list is not that hard to maintain. It should be a simple file with one project name per line. Also ignoring system namespaces like kube* and openshift* should be built into the script and not in the exclusion list

Reply:
As discussed in cabal, we are going for labels. I would recommend a label like 'testing' to make script more generic and reusable and generic where script label indicates project was created for testing.

Filter projects with testing label and delete the old ones.

Reply:
wdyt @feloy

Reply:
I agree with this. Depending when what was discussed with other teams, perhaps also add a label defining the team, to filter on this label too, something similar to: `team=odo&app=testing`

Reply:
Yep, and team name can be an argument to script

Reply:
Do we want to add `app=testing` to the script as well?

Reply:
I guess @feloy is pointing to the testing label I mentioned. The form is irrelevant. The important thing is we only destroy projects that

1. Have a `testing` label. This can be `app=testing`
2. Have a `team=blah` label where `blah` can be passed to the script.
3. Older than X
This assumes that only projects created by tests have the `testing` label

Reply:
Understood, I have made the changes.